File: packages/goals_core/lib/src/book_metadata.dart
import 'package:goals_types/goals_types.dart';

const BOOK_SECTION_ENTRY_TYPE = 'bS';

class BookSectionLogEntry extends GoalLogEntry {
  static const FIRST_VERSION = 1;
  static const START_PAGE_JSON_KEY = 'sP';
  static const END_PAGE_JSON_KEY = 'eP';
  static const START_PAGE_FORMAT_JSON_KEY = 'sPF';
  static const END_PAGE_FORMAT_JSON_KEY = 'ePF';
  static const BOOK_GOAL_ID_JSON_KEY = 'bI';

  final int startPage;
  final int? endPage;
  final String startPageFormat;
  final String? endPageFormat;
  final String? bookGoalId;

  const BookSectionLogEntry({
    required super.id,
    required super.creationTime,
    required this.startPage,
    this.endPage,
    this.startPageFormat = 'arabic',
    this.endPageFormat,
    this.bookGoalId,
    super.path,
  });

  @override
  List<Object?> get props => [
        id,
        creationTime,
        startPage,
        endPage,
        startPageFormat,
        endPageFormat,
        bookGoalId
      ];

  static BookSectionLogEntry fromJsonMap(dynamic json, int? version) {
    if (json is! Map) {
      throw Exception('Invalid data: $json is not a map');
    }
    return BookSectionLogEntry(
      id: json[GoalLogEntry.ID_JSON_KEY],
      creationTime: DateTime.fromMillisecondsSinceEpoch(
          json[GoalLogEntry.CREATION_TIME_JSON_KEY]!),
      startPage: json[START_PAGE_JSON_KEY] is int
          ? json[START_PAGE_JSON_KEY]
          : int.parse(json[START_PAGE_JSON_KEY].toString()),
      endPage: json[END_PAGE_JSON_KEY],
      startPageFormat:
          json[START_PAGE_FORMAT_JSON_KEY] ?? json['pF'] ?? 'arabic',
      endPageFormat: json[END_PAGE_FORMAT_JSON_KEY],
      bookGoalId: json[BOOK_GOAL_ID_JSON_KEY],
      path: json[GoalLogEntry.PATH_JSON_KEY]?.cast<String>(),
    );
  }

  @override
  Map<String, dynamic> toJsonMap() {
    return {
      ...super.toJsonMap(),
      GoalLogEntry.TYPE_JSON_KEY: BOOK_SECTION_ENTRY_TYPE,
      START_PAGE_JSON_KEY: startPage,
      if (endPage != null) END_PAGE_JSON_KEY: endPage,
      START_PAGE_FORMAT_JSON_KEY: startPageFormat,
      if (endPageFormat != null) END_PAGE_FORMAT_JSON_KEY: endPageFormat,
      if (bookGoalId != null) BOOK_GOAL_ID_JSON_KEY: bookGoalId,
    };
  }
}

class BookLogEntryModule implements LogEntryModule {
  @override
  GoalLogEntry? fromJsonMap(dynamic json, int? version) {
    if (json is Map &&
        json[GoalLogEntry.TYPE_JSON_KEY] == BOOK_SECTION_ENTRY_TYPE) {
      return BookSectionLogEntry.fromJsonMap(json, version);
    }
    return null;
  }
}

File: packages/goals-core/src/book-metadata.ts
import { GoalLogEntry, WireGoalLogEntry } from "@thkp-eng/goals-types";

export const BOOK_SECTION_ENTRY_TYPE = "bS";

// Mimic the Dart constants
export const FIRST_VERSION = 1;
export const START_PAGE_JSON_KEY = "sP";
export const END_PAGE_JSON_KEY = "eP";
export const START_PAGE_FORMAT_JSON_KEY = "sPF";
export const END_PAGE_FORMAT_JSON_KEY = "ePF";
export const BOOK_GOAL_ID_JSON_KEY = "bI";

/**
 * Wire format for Book Section Log Entry
 */
export interface WireBookSectionLogEntry extends WireGoalLogEntry {
  t: typeof BOOK_SECTION_ENTRY_TYPE;
  sP: number | string;
  eP?: number | string | null;
  sPF?: string | null;
  ePF?: string | null;
  bI?: string | null;
}

/**
 * Pretty format for Book Section Log Entry
 */
export interface BookSectionLogEntry extends GoalLogEntry {
  type: typeof BOOK_SECTION_ENTRY_TYPE;
  startPage: number;
  endPage?: number | null;
  startPageFormat: string;
  endPageFormat?: string | null;
  bookGoalId?: string | null;
}

/**
 * Mimic the Dart fromJsonMap equivalent
 */
export function bookSectionLogEntryFromWire(
  wire: any,
): BookSectionLogEntry | null {
  if (wire?.t === BOOK_SECTION_ENTRY_TYPE || wire?.type === BOOK_SECTION_ENTRY_TYPE) {
    const sP = wire[START_PAGE_JSON_KEY] ?? wire.sP;
    return {
      id: wire.i || wire.id,
      creationTime: wire.cT || wire.creationTime,
      path: wire.p || wire.path,
      type: BOOK_SECTION_ENTRY_TYPE,
      startPage: typeof sP === "number" ? sP : parseInt(sP, 10),
      endPage: wire[END_PAGE_JSON_KEY] ?? wire.eP,
      startPageFormat:
        wire[START_PAGE_FORMAT_JSON_KEY] ??
        wire.sPF ??
        wire.pF ?? // Fallback to old key as in Dart
        "arabic",
      endPageFormat: wire[END_PAGE_FORMAT_JSON_KEY] ?? wire.ePF,
      bookGoalId: wire[BOOK_GOAL_ID_JSON_KEY] ?? wire.bI,
    };
  }
  return null;
}

/**
 * Mimic the Dart toJsonMap equivalent
 */
export function bookSectionLogEntryToWire(
  pretty: BookSectionLogEntry,
): WireBookSectionLogEntry {
  const wire: any = {
    i: pretty.id,
    cT: pretty.creationTime,
    t: BOOK_SECTION_ENTRY_TYPE,
    [START_PAGE_JSON_KEY]: pretty.startPage,
    [START_PAGE_FORMAT_JSON_KEY]: pretty.startPageFormat ?? "arabic",
  };

  if (pretty.path) wire.p = pretty.path;
  if (pretty.endPage !== undefined && pretty.endPage !== null) {
    wire[END_PAGE_JSON_KEY] = pretty.endPage;
  }
  if (pretty.endPageFormat !== undefined && pretty.endPageFormat !== null) {
    wire[END_PAGE_FORMAT_JSON_KEY] = pretty.endPageFormat;
  }
  if (pretty.bookGoalId !== undefined && pretty.bookGoalId !== null) {
    wire[BOOK_GOAL_ID_JSON_KEY] = pretty.bookGoalId;
  }

  return wire as WireBookSectionLogEntry;
}
